### PR TITLE
Fix db migration errors on sqlite (and potentially others)

### PR DESCRIPTION
**Pull request by rjarry**
When a build is cancelled while a step is starting up. The stepid has not been initialized yet. Before #4539, there was no integrity constraints on the stepid column which may have caused "invalid" rows to be created.

This may cause migration number 52 to fail since the constraint is violated.

In case such offending rows are found during the migration, issue a warning and delete them since they are indeed invalid.

Fixes: #4711 

## Contributor Checklist:

* [x] I have updated the unit tests
* [x] I have created a file in the `master/buildbot/newsfragment` directory (and read the `README.txt` in that directory)

### CODE REVIEW

File: master/buildbot/process/buildstep.py [hunk 1]
```diff
@@ -789,6 +789,8 @@ def getWorkerName(self):
         return self.build.getWorkerName()
 
     def addLog(self, name, type='s', logEncoding=None):
+        if self.stepid is None:
```
Comment (tardyp) [inline master/buildbot/process/buildstep.py]: looks like unrelated

Comment (rjarry) [reply master/buildbot/process/buildstep.py]: This is the root cause of the problem. 

Since stepid is None until it is actually added to the db, if a build is cancelled before the build step is properly initialized, there may be a call to addLog while stepid is still None.

Now that there are NOT NULL constraints on the logs.stepid column, we should not experience db integrity problems. However, this would cause errors to be risen.

Comment (rjarry) [reply master/buildbot/process/buildstep.py]: Please note that I took care of putting this in a separate commit.

Comment (p12tic) [reply master/buildbot/process/buildstep.py]: A better PR description would have prevented this misunderstanding :-)

Comment (rjarry) [reply master/buildbot/process/buildstep.py]: You're being polite :-)

```bash
| sed 's/better //'
```

PR description updated.

**Comment by rjarry** [comment]
> Code looks good. Could we cover the behavior of not adding logs to DB if stepid is None with unit tests?

Yes that would be better. I did not have much time to work on this (hence the absence of PR description :-))

I'll try to do this soon.

**Comment by p12tic** [comment]
Thanks!

File: master/buildbot/process/buildstep.py (hunk 1 above)
Comment (p12tic) [reply master/buildbot/process/buildstep.py]: @tardyp: Could you re-review the PR? Then I'll merge it as soon as we have tests.

**Comment by tardyp** [comment]
As this fix some migration issue. I am merging it.
